url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+     * @param inst ARM 命令
+     * @return アドレスの文字列表記
+     */
+    public String getAddrMode2Name(InstructionARM inst) {
+        boolean i = inst.getIBit();
+        boolean p = inst.getBit(24);
+        boolean u = inst.getBit(23);
+        //boolean b = inst.getBit(22);
+        boolean w = inst.getBit(21);
+        int rn = inst.getRnField();
+        int shift_imm = inst.getField(7, 5);
+        int shift = inst.getField(5, 2);
+        //int rm = inst.getRmField();
+        String strOffset;
+
+        if (!i) {
+            //12bits イミディエートオフセット
+            //I ビットの意味がデータ処理命令と逆なので注意！
+            return getAddrMode2ImmName(inst);
+        } else if (shift_imm == 0 && shift == 0) {
+            //レジスタオフセット/インデクス
+            strOffset = getAddrMode2RegName(inst);
+        } else {
+            //スケーリング済みレジスタオフセット/インデクス
+            strOffset = getAddrMode2ScaledName(inst);
+        }
+
+        if (p && !w) {
+            //オフセット
+            return String.format("[%s, %s%s]",
+                    getRegName(rn), (u) ? "" : "-",
+                    strOffset);
+        } else if (p && w) {
+            //プリインデクス
+            return String.format("[%s, %s%s]!",
+                    getRegName(rn), (u) ? "" : "-",
+                    strOffset);
+        } else if (!p) {
+            //ポストインデクス
+            return String.format("[%s], %s%s",
+                    getRegName(rn), (u) ? "" : "-",
+                    strOffset);
+        } else {
+            throw new IllegalArgumentException("Illegal P,W bits " +
+                    String.format("p:%b, w:%b.", p, w));
+        }
+    }
+
+    /**
+     * アドレシングモード 2 - ワードまたは符号無しバイトロード/ストア、
+     * 12bits イミディエートオフセットアドレスの、
+     * 文字列表記を取得します。
+     *
+     * 下記の特殊記法を採用しているため、個別に処理しています。
+     * 正負符号の前に # 記号が入る。
+     * 角括弧の後に 16進数表記が入る。
+     *
+     * @param inst ARM 命令
+     * @return イミディエートオフセットアドレスの文字列表記
+     */
+    public String getAddrMode2ImmName(InstructionARM inst) {
+        boolean p = inst.getBit(24);
+        boolean u = inst.getBit(23);
+        //boolean b = inst.getBit(22);
+        boolean w = inst.getBit(21);
+        int rn = inst.getRnField();
+        int offset12 = inst.getField(0, 12);
+
+        if (p && !w) {
+            //イミディエートオフセット
+            return String.format("[%s, #%s%d]    ; 0x%x",
+                    getRegName(rn), (u) ? "" : "-",
+                    offset12, offset12);
+        } else if (p && w) {
+            //プリインデクスイミディエート
+            return String.format("[%s, #%s%d]!    ; 0x%x",
+                    getRegName(rn), (u) ? "" : "-",
+                    offset12, offset12);
+        } else if (!p) {
+            //ポストインデクスイミディエート
+            return String.format("[%s], #%s%d    ; 0x%x",
+                    getRegName(rn), (u) ? "" : "-",
+                    offset12, offset12);
+        } else {
+            throw new IllegalArgumentException("Illegal P,W bits " +
+                    String.format("p:%b, w:%b.", p, w));
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 810:828, 947:965.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.